### PR TITLE
Markdown documentation for `DecisionTree`, `DecisionTreeRegressor`, and `RandomForest`

### PR DESCRIPTION
**Pull request by rcurtin**
This follows the discussion in https://github.com/mlpack/mlpack.org/issues/59.

I wrote up a very first attempt at simple documentation for the `DecisionTree<>` class that follows the style of the Armadillo documentation and the ensmallen documentation.  The general idea is to provide a very simple set of instructions so that a user can figure out how to quickly create, train, and predict with a decision tree.  Further, I also included enough breadcrumbs to help more advanced users.  (I haven't finished the template parameters section.  I hope to keep it short.)

I wanted to get some feedback---especially from @conradsnicta ---on how this looks and reads.  I haven't yet figured out quite how a user will navigate down to find this particular documentation, but I'm definitely open to ideas and suggestions.  Some general thoughts I kept in mind here:

 * Template parameters should be mostly hidden.  (I think actually we should rename `DecisionTree<>` to `DecisionTreeType<>` then typedef `DecisionTree` to the default template parameters, so that potentially scary `<>`s don't need to appear in simple user code.)

 * Wherever things are a little more complicated than meets the eye, I try to link to further documentation.  So, e.g., any time we have a matrix parameter, I link to matrix documentation so that users can find out (if they don't already know) that their observations should be columns and dimensions should be rows, etc.

 * I did my best to barely bother with any words, instead assuming the user will ignore words and focus on code.  That's... very hard for me. :)

 * I highlighted more advanced functionality as free-form bullet points, instead of a disorganized method list like Doxygen might give.

Once we can generally agree on an idea of how this type of documentation should look, I can use this as a template for other algorithms and classes inside the codebase.

### CODE REVIEW

**Comment by rcurtin** [comment]
(CC: @shrit @zoq)

File: doc/user/methods/decision_tree.md [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+
+ * `tree.Train(`_`data`_`, `_`labels`_`, `_`numClasses`_`, `_`weights`_`)` (train on weighted numerical-only data)
+ * `tree.Train(`_`data`_`, `_`labels`_`, `_`numClasses`_`, `_`weights`_`, `_`minimumLeafSize`_`, `_`minimumGainSplit`_`, `_`maximumDepth`_`)` (train on weighted numerical-only data with hyperparameters)
+
+ * `tree.Train(`_`data`_`, `_`datasetInfo`_`, `_`labels`_`, `_`numClasses`_`)` (train on mixed categorical data)
+ * `tree.Train(`_`data`_`, `_`datasetInfo`_`, `_`labels`_`, `_`numClasses`_`, `_`minimumLeafSize`_`, `_`minimumGainSplit`_`, `_`maximumDepth`_`)` (train on mixed categorical data with hyperparameters)
+
+ * `tree.Train(`_`data`_`, `_`datasetInfo`_`, `_`labels`_`, `_`numClasses`_`, `_`weights`_`)` (train on weighted mixed categorical data)
+ * `tree.Train(`_`data`_`, `_`datasetInfo`_`, `_`labels`_`, `_`numClasses`_`,
+   `_`weights`_`, `_`minimumLeafSize`_`, `_`minimumGainSplit`_`, `_`maximumDepth`_`)` (train on weighted mixed categorical data with hyperparameters)
```
Comment (shrit) [inline doc/user/methods/decision_tree.md:58]: do we keep the declaration or replace them with examples directly?

**Review by shrit — COMMENTED** [review]
I like these, they are much better than before. However, I can not visualize them, but I am trying to imagine what this will look like.
I left one comment regarding the function declaration maybe it will be easier to have only usage examples for each one of the overloads instead of the function declaration, and remove the simple examples section. Since the user is expecting how to use them directly.

**Comment by conradsnicta** [comment]
@rcurtin I have a few work deadlines to meet, so will look at this properly in February. Apologies.

**Comment by rcurtin** [comment]
> @rcurtin I have a few work deadlines to meet, so will look at this properly in February. Apologies.

No worries---my main focus is on Bandicoot right now, so I'm not sure when I'll really be able to dig in on this yet.

File: doc/user/methods/decision_tree.md (hunk 1 above)
Comment (rcurtin) [reply doc/user/methods/decision_tree.md:58]: Hmm, it's a good point.  I was hoping to keep all the different overloads close together in the documentation, like for Armadillo and ensmallen.  Also, I'm not sure that we need an example for every single constructor---the difference between the first two, for instance, is trivial.  I'm trying to think about how I usually use Armadillo documentation: I take a look at the function signatures, and then take a guess at how I'd use that in my code, and usually, the guess is right.  If not, then I would go to the examples or further documentation to see if I missed something.  I'm not sure if that's how most people look at Armadillo documentation, but that's at least what I do...

**Comment by conradsnicta** [comment]
@rcurtin Had a look at the new docs by pasting them into https://markdownlivepreview.com (also tried https://stackedit.io/app# and https://dillinger.io/).

In the table explaining the variables, I suggest swapping the order of _type_ and _name_ columns. This would make looking up the variables names easier.

(There is a minor rendering issue with the the table; removing the TODO comments seems to fix the rendering.)

In the Training section, the "tree" instance comes out of nowhere, which might be confusing. I suggest stating something along these lines: "For an instance of _DecisionTree_ named as _tree_, the following functions for training are available:"

Suggest to avoid the use of "overloads" within "... which has several overloads", as it's not clear to newbies what is meant by "overload" in this context.

Suggest to add a "Classification" section, which explains the functions used for classification.

It would be also very useful to add an "Examples" section (similar to what Armadillo docs have for almost every function) which would show how to concretely use the _DecisionTree_ class within code.  The example itself doesn't need to use real data; random (but otherwise valid) matrices are sufficient.  Short and sweet examples are fine.  Such examples can be copy'n'pasted by users and quickly adapted to their needs, or just to get a feel how the class can be trained and used.

Lastly, I suggest adding a "See Also" section, which has links to relevant info.  For example, the following links would be useful as part of the _DecisionTree_ docs:
* https://en.wikipedia.org/wiki/Decision_tree
* https://en.wikipedia.org/wiki/Decision_tree_learning

The relevant info can be thought of as "extensions to the documentation".

**Comment by rcurtin** [comment]
Finally managed to get back to this.  I went through and worked in all the comments (thanks @conradsnicta and @shrit, very helpful), and filled out the rest of the stub sections that I wanted to add.  A couple quick thoughts:

 * I don't like the Github Markdown styling; honestly, it looks [kind of ugly](https://github.com/rcurtin/mlpack/blob/rf-doc/doc/user/methods/decision_tree.md), but it's at least readable.  I intend to deploy this eventually to the website, where I can do some additional css/processing to clean things up.  I'll get an example of that set up soon.
 
  * I tried to make it clear and simple that you can use different element types (e.g. you can train a tree with `arma::fmat`), but I don't know how good of a job I did.  Happy to take any feedback.
 
  * The template section is long and most users won't deal with it, so I left it at the bottom, but I think it's important to include, since we advertise mlpack as very flexible.

**Comment by rcurtin** [comment]
After some minor additional cleanup, here's Kramdown-generated HTML with some custom CSS styling applied:

http://ratml.org/misc/mlpack-markdown-doc/decision_tree.html

It displays okay (but not quite as readable) as Github-rendered Markdown, which was my goal.

**Comment by rcurtin** [comment]
I went ahead and added similar documentation for `DecisionTreeRegressor` and `RandomForest`.  All of the code examples are tested, and compile and run as expected.  (It would be cool to automate that, but I'm not yet sure how feasible---or worth the effort---that is.)

Here are sample pages with kramdown-compiled documentation and some CSS:

 * https://www.ratml.org/misc/mlpack-markdown-doc/decision_tree.html
 * https://www.ratml.org/misc/mlpack-markdown-doc/decision_tree_regressor.html
 * https://www.ratml.org/misc/mlpack-markdown-doc/random_forest.html

Note that the `RandomForest` documentation has an extra monkey wrench: we use different template parameters to provide an `ExtraTrees` typedef.  I tried to cover this with a `Variants` section, but I'm open to other suggestions too.

I won't update this PR more after this---I'll continue to document other methods, but open different PRs for them (or groups of them).

**Comment by rcurtin** [comment]
I gathered some feedback and applied it.  The links in the post above are updated with example renderings:

 * https://www.ratml.org/misc/mlpack-markdown-doc/decision_tree.html
 * https://www.ratml.org/misc/mlpack-markdown-doc/decision_tree_regressor.html
 * https://www.ratml.org/misc/mlpack-markdown-doc/random_forest.html

A summary of the changes:

 * Reduced a whole lot of redundant text.
 * The initial example is now a full copy-pastable working example, and has links to other examples directly below.
 * Removed the `Variants` section from the front of the random forest documentation; instead, this is replaced with a short blurb indicating that you can use `ExtraTrees` instead of `RandomForest`, and links to a section in the advanced template parameter documentation that details it.

**Review by zoq — APPROVED** [review]

**Comment by rcurtin** [comment]
I'm going to go ahead and merge this tomorrow, but if there are any final comments from @shrit or @coatless or others, happy to handle them (or we can always do it in a follow-up PR).

**Comment by coatless** [comment]
@rcurtin in we go 👍

**Pull request merged**